datafusion/core/src/datasource/flight/sql.rs

```diff
@@ -0,0 +1,475 @@
+use arrow_flight::sql::{CommandStatementQuery, ProstMessageExt};
+use arrow_flight::{FlightDescriptor, FlightInfo, HandshakeRequest, HandshakeResponse};
+use arrow_schema::ArrowError;
+use async_trait::async_trait;
+use base64::prelude::BASE64_STANDARD;
+use base64::Engine;
+use bytes::Bytes;
+use futures::{stream, TryStreamExt};
+use prost::Message;
+use tonic::metadata::{AsciiMetadataKey, MetadataMap};
+use tonic::transport::Channel;
+use tonic::IntoRequest;
+
+use crate::datasource::flight::{FlightDriver, FlightMetadata};
+
+/// Default Flight SQL driver. Requires a `flight.sql.query` to be passed as a table option.
+/// If `flight.sql.username` (and optionally `flight.sql.password`) are passed,
+/// will perform the `Handshake` using basic authentication.
+/// Any additional headers can be passed as table options using the `flight.sql.header.` prefix.
+///
+/// A [crate::datasource::flight::FlightTableFactory] using this driver is registered
+/// with the default `SessionContext` under the name `FLIGHT_SQL`.
+#[derive(Clone, Debug, Default)]
+pub struct FlightSqlDriver {}
+
+#[async_trait]
+impl FlightDriver for FlightSqlDriver {
+    async fn metadata(
+        &self,
+        channel: Channel,
+        options: &HashMap<String, String>,
+    ) -> Result<FlightMetadata> {
+        let mut client = FlightSqlClient::new(channel);
+        let headers = options.iter().filter_map(|(key, value)| {
+            key.strip_prefix("flight.sql.header.")
+                .map(|header_name| (header_name, value))
+        });
+        for header in headers {
+            client.set_header(header.0, header.1)
+        }
+        if let Some(username) = options.get("flight.sql.username") {
+            let default_password = "".to_string();
+            let password = options
+                .get("flight.sql.password")
+                .unwrap_or(&default_password);
+            _ = client.handshake(username, password).await?;
+        }
+        let info = client
+            .execute(options["flight.sql.query"].clone(), None)
+            .await?;
+        let mut grpc_metadata = MetadataMap::new();
+        if let Some(token) = client.token {
+            grpc_metadata.insert(
+                "authorization",
+                format!("Bearer {}", token).parse().unwrap(),
+            );
+        }
+        FlightMetadata::try_new(info, grpc_metadata)
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////
+// Shameless copy/paste from arrow-flight FlightSqlServiceClient
+// This is only needed in order to access the bearer token received
+// during handshake, as the standard client does not expose this information.
+// The bearer token has to be passed to the clients that perform
+// the DoGet operation, since Dremio, Ballista and possibly others
+// expect the bearer token they produce with the handshake response
+// to be set on all subsequent requests, including DoGet.
```

Comment:
TODO: Remove this "stolen" client implementation and submit a PR to arrow-rs to make the token visible in the official client.

Reply:
https://github.com/apache/arrow-rs/issues/6253

Reply:
https://github.com/apache/arrow-rs/pull/6254

Reply:
@alamb on average, my PRs are of reasonable size. :)

Reply:
I appreciate it @ccciudatu  -- I am sorry I haven't had a chance to comment more thoroughly on this PR yet...

Reply:
No worries, @alamb! The fact that you are even considering reviewing such a huge patch makes you one of the nicest people I know. 🙃

Reply:
that is nice of you to say -- the whole point of doing this in open source is to get other people to contribute things and a key way to do that is to actually look at their proposals 😆